File: flink_scala/src/main/scala/atguigu/base/UvCountByWindowExample.scala
package atguigu.base

import org.apache.flink.api.scala.createTypeInformation
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.streaming.api.scala.function.ProcessWindowFunction
import org.apache.flink.streaming.api.windowing.assigners.TumblingEventTimeWindows
import org.apache.flink.streaming.api.windowing.time.Time
import org.apache.flink.streaming.api.windowing.windows.TimeWindow
import org.apache.flink.util.Collector

import java.sql.Timestamp

object UvCountByWindowExample {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)
    env.addSource(new ClickSource).assignAscendingTimestamps(_.timestamp1)
      // 为所有数据都指定同一个 key，可以将所有数据都发送到同一个分区
      .keyBy(_ => "key")
      .window(TumblingEventTimeWindows.of(Time.seconds(10)))
      .process(new UvCountByWindow)
      .print()
    env.execute()
  }

  // 自定义窗口处理函数
  /*
    ProcessWindowFunction:是窗口函数中的全窗口函数中的处理窗口函数
    全窗口聚合函数需要先收集窗 口中的数据，并在内部缓存起来，等到窗口要输出结果的时候再取出数据进行计算
    elements：就是窗口中缓存的数据
   */
  class UvCountByWindow extends ProcessWindowFunction[Event, String, String, TimeWindow] {
    override def process(key: String, context: Context, elements: Iterable[Event], out: Collector[String]): Unit = {
      // 初始化一个 Set 数据结构，用来对用户名进行去重
      var userSet = Set[String]()
      // 将所有用户名进行去重
      elements.foreach(userSet += _.user1)
      // 结合窗口信息，包装输出内容
      val windowStart = context.window.getStart
      val windowEnd = context.window.getEnd
      out.collect(" 窗 口 ： " + new Timestamp(windowStart) + "~" + new Timestamp(windowEnd) + "的独立访客数量是：" + userSet.size)
    }
  }
}

File: flink_java/src/main/java/flink01/chapter06/UvCountByWindowExample.java
package flink01.chapter06;

import flink01.chapter05.ClickSource;
import flink01.chapter05.Event;
import org.apache.flink.api.common.eventtime.SerializableTimestampAssigner;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.windowing.ProcessWindowFunction;
import org.apache.flink.streaming.api.windowing.assigners.TumblingEventTimeWindows;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.streaming.api.windowing.windows.TimeWindow;
import org.apache.flink.util.Collector;

import java.sql.Timestamp;
import java.time.Duration;
import java.util.HashSet;

public class UvCountByWindowExample {

    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);
        SingleOutputStreamOperator<Event> stream = env.addSource(new ClickSource())
                                                      .assignTimestampsAndWatermarks(WatermarkStrategy.<Event>forBoundedOutOfOrderness(Duration.ZERO)
                                                                                                      .withTimestampAssigner((SerializableTimestampAssigner<Event>) (element, recordTimestamp) -> element.timestamp));
        // 将数据全部发往同一分区，按窗口统计UV
        stream.keyBy(data -> true)
              .window(TumblingEventTimeWindows.of(Time.seconds(10)))
              .process(new ProcessWindowFunction<Event, String, Boolean, TimeWindow>() {
                  @Override
                  public void process(Boolean aBoolean, ProcessWindowFunction<Event, String, Boolean, TimeWindow>.Context context, Iterable<Event> elements, Collector<String> out) {
                      HashSet<Object> userSet = new HashSet<>();
                      for (Event element : elements) {
                          userSet.add(element.user);
                      }
                      // 结合窗口信息，包装输出内容
                      long start = context.window()
                                          .getStart();
                      long end = context.window()
                                        .getEnd();
                      out.collect("窗口: " + new Timestamp(start) + " ~ " + new Timestamp(end) + " 的独立访客数量是：" + userSet.size());
                  }
              })
              .print();
        env.execute();
    }
}
